Adds test for authentication with an invalid token

jkakavas opened the pull request:
Ensure that we return 401 when an invalid token is presented as
the Bearer token

Relates: #38866

jkakavas committed:
Adds test for authentication with an invalid token

Ensure that we return 401 when an invalid token is presented as
the Bearer token

Relates: #38866

elasticmachine commented:
Pinging @elastic/es-security

albertzaharovits commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java:
```diff
@@ -268,6 +268,23 @@ public void testRefreshingToken() throws IOException {
+            RequestOptions.DEFAULT.toBuilder().addHeader("Authorization", "Bearer " + randomAlphaOfLengthBetween(0, 128)).build();
+        ElasticsearchStatusException e = expectThrows(ElasticsearchStatusException.class,
+            () -> restClient.security().authenticate(wrongAuthOptions));
+        assertEquals(RestStatus.UNAUTHORIZED, e.status());
```
I was expecting you'd invalidate the created token and try another time with it, rather than creating a new random string.

We have a few formats (lengths) for tokens. If you wish to test some/most of them, I would suggest we craft them on purpose, so that we cover the branches in `decodeToken`. I think a long random string is insufficient to cover for "invalid" tokens, generally.

If you only test the REST response code, which I believe you are aiming for, a random string is enough (no need to create another token).

albertzaharovits reviewed (commented):
Are we testing "invalid" tokens generally, or just that the HTTP response is 401 and not 500?

jkakavas commented:
> Are we testing "invalid" tokens generally, or just that the HTTP response is 401 and not 500?

The latter. 

> I was expecting you'd invalidate the created token and try another time with it, rather than using a random string.

The idea is to test that if we get an access token that we don't recognise/can't validate, we respond with the proper HTTP status code ( 401 instead of 500 - as we used to do in the past ) 

> We have a few formats (lengths) for tokens. If you wish to test some/most of them, I would suggest we craft them on purpose, so that we cover the branches in decodeToken.

No objection to do this , but I'd probably fit these tests in `TokenServiceTests`

albertzaharovits commented:
Then in this case I would assert the 401 in all the other tests that invalidate/expire tokens, rather thank creating a separate test for it, WDYT?

The problem with a standalone test with a random string is that it is to flaky given the multitude of formats we have. And hardcoding one token format is not very future proof.

jkakavas commented:
`testClientCredentialsGrant` already does and I could do the same for the ones that invalidate all tokens for a user/realm. But what you propose is to test with access tokens that we have invalidated or have been expired. What I'd like us to test is that we return a 401 with an arbitrary string that is not an access token we have created previously and then expired or invalidated. 

> The problem with a standalone test with a random string is that it is to flaky given the multitude of formats we have.

This is what I'm planning to test with this one. No matter how many formats we have, an 8.0.0 or > 7.2.0 node, should not return a 500 no matter what the token value is. I'm up to adding explicit tests for the different formats in order to build our certainty, but covering that expired/invalidated tokens return 401 is not enough I think.

albertzaharovits commented:
Okay, then provided that invalidated/expired tokens are already tested for 401, what needs to be tested is a bogus/missing token, one that never existed. As I said there are many formats, and we prepend a version string, so I am not convinced that a random string is enough to cover this case.

Given that tokens are short-lived, I believe a missing token in the last format is enough. To make the test future proof it's porb better to generate the token then remove it (delete `.security-tokens-7`), rather than hardcode a value for it in the test.

jkakavas commented:
> As I said there are many formats, and we prepend a version string, so I am not convinced that a random string is enough to cover this case. Given that tokens are short-lived, I believe a missing token in the last format is enough.

The many formats and the version prepending are internal documentation details. What we need to be testing here, is:  

 - Is the token _correct_ ? : Then we we accept this as authentication information and proceed to perform authZ checks respond.
- Is the token _not correct_ ? : We don't care why this was not correct. Is it because it was invalid, already expired and deleted, already expired but not yet deleted ? Never created in the first place ?  We should respond with 401. Not with a 500.  

Only dealing with tokens we have created ourselves and match the format we expect doesn't cover for the case where a client sends a truncated token, a malformed one, or a token from another Authorization server. The reason why this test deals only with the truncated/malformed/other AS token is that I felt the expired/invalidated where covered in other existing tests. I have no objection to making the test more thorough to explicitly check for other formats but this will never cater for the strange client that will send `Bearer thisisnotatokenthatweexpectbutweshouldstillreturna401andnotbarfa500asifsomethingwentwrong` in the Authorization header. Pre 7.2, we would throw a 500 for this, I believe this is wrong and this small test case aims to ensure that this kind of behavior won't be reintroduced at some point.


p.s.:  We have a known bug, in that we return 500 for tokens that we have deleted : see https://github.com/elastic/elasticsearch/issues/38866

albertzaharovits commented:
Understood.
So, if the goal is to test that **every** malformed token returns 401, I think it needs to take into account the formats we use, so that we make sure that every else branch in the `TokenService#decode` results in a 401. A random string won't work because some failure cases are very hard to trip (eg missing decryption key) compared with the IOException for the prepended version.
But these failure cases are **hard to test in an IntegTest** (they need specially crafted inputs) and are not future proof. **I think the same cases should be tested in `TokenServiceTests`** and some of them are (eg missing key).

jkakavas commented:
> So, if the goal is to test that every malformed token returns 401, I think it needs to take into account the formats we use, so that we make sure that every else branch in the TokenService#decode results in a 401.

Agreed. 

>  A random string won't work because some failure cases are very hard to trip (eg missing decryption key) compared with the IOException for the prepended version.

I'm not claiming we should _only_ test with random strings, but I think that using random strings as part of the tests allows us to test branches of the TokenService#decode and behavior that sticking only to known/expected formats won't. 

> But these failure cases are hard to test in an IntegTest (they need specially crafted inputs) and are not future proof. I think the same cases should be tested in TokenServiceTests and some of them are (eg missing key).

Agreed (https://github.com/elastic/elasticsearch/pull/45138#issuecomment-518270530): 

> No objection to do this , but I'd probably fit these tests in TokenServiceTests

jkakavas commented:
So, would you prefer if I close this and reopen one at some point that handles all cases we can think of with tests in here and TokenServiveTests ?

albertzaharovits commented:
Sounds good to me.
